cmd/restic/cmd_prune.go

```diff
@@ -87,229 +136,433 @@ func runPrune(gopts GlobalOptions) error {
+		Verbosef("deleting unused index files...\n")
+		DeleteFiles(gopts, true, repo, obsoleteIndexes, restic.IndexFile)
 	}
 
-	Verbosef("will delete %d packs and rewrite %d packs, this frees %s\n",
-		len(removePacks), len(rewritePacks), formatBytes(uint64(removeBytes)))
+	// unreferenced packs can be safely deleted first
+	if len(removePacksFirst) != 0 {
+		Verbosef("deleting unreferenced data files...\n")
+		DeleteFiles(gopts, true, repo, removePacksFirst, restic.DataFile)
+	}
 
-	var obsoletePacks restic.IDSet
-	if len(rewritePacks) != 0 {
-		bar = newProgressMax(!gopts.Quiet, uint64(len(rewritePacks)), "packs rewritten")
+	if len(repackPacks) != 0 {
+		Verbosef("repacking data files...\n")
+		bar := newProgressMax(!gopts.Quiet, uint64(len(repackPacks)), "data files repacked")
 		bar.Start()
-		obsoletePacks, err = repository.Repack(ctx, repo, rewritePacks, usedBlobs, bar)
+		// TODO in Repack:  - Parallelize repacking
+		//                  - Make use of blobs stored multiple times (e.g. if SHA doesn't match)
+		_, err := repository.Repack(ctx, repo, repackPacks, keepBlobs, bar)
+		bar.Done()
 		if err != nil {
 			return err
 		}
-		bar.Done()
+		// Also remove repacked packs
+		removePacks.Merge(repackPacks)
 	}
 
-	removePacks.Merge(obsoletePacks)
+	if len(removePacks) != 0 {
+		Verbosef("rebuilding index files...\n")
 
-	if err = rebuildIndex(ctx, repo, removePacks); err != nil {
-		return err
-	}
+		// Call RebuildIndex: rebuilds the index from the already loaded in-memory index.
+		// TODO in RebuildIndex: - Parallelize rebuilding
+		obsoleteIndexes, err := (repo.Index()).(*repository.MasterIndex).
+			RebuildIndex(ctx, repo, removePacks)
+		if err != nil {
+			return err
+		}
 
-	if len(removePacks) != 0 {
-		bar = newProgressMax(!gopts.Quiet, uint64(len(removePacks)), "packs deleted")
-		bar.Start()
-		for packID := range removePacks {
-			h := restic.Handle{Type: restic.DataFile, Name: packID.String()}
-			err = repo.Backend().Remove(ctx, h)
-			if err != nil {
-				Warnf("unable to remove file %v from the repository\n", packID.Str())
-			}
-			bar.Report(restic.Stat{Blobs: 1})
+		Verbosef("deleting obsolete index files...\n")
+		err = DeleteFiles(gopts, false, repo, obsoleteIndexes, restic.IndexFile)
+		if err != nil {
+			return err
 		}
-		bar.Done()
+
+		Verbosef("deleting obsolete data files...\n")
+		DeleteFiles(gopts, true, repo, removePacks, restic.DataFile)
 	}
 
-	Verbosef("done\n")
+	Verbosef("done.\n")
 	return nil
 }
+
+const numDeleteWorkers = 8
+
+// DeleteFiles deletes the given fileList of fileType in parallel
+// if ignoreError=true, it will print a warning if there was an error, else it will abort.
+func DeleteFiles(gopts GlobalOptions, ignoreError bool, repo restic.Repository, fileList restic.IDSet, fileType restic.FileType) error {
+	totalCount := len(fileList)
+	fileHandles := make(chan restic.Handle)
+	go func() {
+		for id := range fileList {
+			fileHandles <- restic.Handle{Type: fileType, Name: id.String()}
```

Comment:
Nitpicking, but you could construct the Handle in the consumer for a tiny bit more parallelism.

Reply:
is changed.

I also realized that it might be a good idea to use this parallel deletion also within `restic forget` to delete snapshot files in parallel.